From d2f0739b617753a3b1d6c702b531f5231d5c7ba2 Mon Sep 17 00:00:00 2001
From: Holger Hans Peter Freyther <holger@freyther.de>
Date: Sat, 16 Mar 2024 20:47:20 +0800
Subject: [PATCH] genbc: Deal with multiple definition errors

genbc uses two bison-generated parsers and one lexer.  Give the parser
globals private names and share yylval explicitly instead of relying on
the linker's former common-symbol behavior.
---
 libgst/genbc-decl.y | 3 +++
 libgst/genbc.c      | 6 ++++++
 libgst/genbc.h      | 1 +
 3 files changed, 10 insertions(+)

diff --git a/libgst/genbc-decl.y b/libgst/genbc-decl.y
index 7c73941d..9d0df7b5 100644
--- a/libgst/genbc-decl.y
+++ b/libgst/genbc-decl.y
@@ -56,6 +56,9 @@
 #include "genbc.h"
 #include "avltrees.h"
 
+#define yynerrs decl_yynerrs
+#define yychar decl_yychar
+#define yylval (*genbc_yylval())
 #define yyparse decl_yyparse
 #define yydebug decl_yydebug
 #define YYERROR_VERBOSE
diff --git a/libgst/genbc.c b/libgst/genbc.c
index cae5f6d3..e6a8042d 100644
--- a/libgst/genbc.c
+++ b/libgst/genbc.c
@@ -54,6 +54,12 @@
 int errors = 0;
 const char *current_file;
 
+YYSTYPE*
+genbc_yylval ()
+{
+  return &yylval;
+}
+
 void
 yyprint (FILE *file, int type, YYSTYPE yylval)
 {
diff --git a/libgst/genbc.h b/libgst/genbc.h
index 44501c8e..6eab493e 100644
--- a/libgst/genbc.h
+++ b/libgst/genbc.h
@@ -78,6 +78,7 @@ extern int impl_yyparse ();
 extern int impl_yydebug;
 
 /* genbc.c declarations */
+extern YYSTYPE *genbc_yylval ();
 extern const char *current_file;
 extern void yyprint (FILE *file, int type, YYSTYPE yylval);
 extern void yyerror (const char *s);
-- 
2.51.0
